### DISCUSSION TITLE
distribute private package to my customers - distribution license now available

### DISCUSSION THREAD

**Discussion by isaac-gofa** [onepub-dev/onepub#285] [Q&A] [answered]
I see in the documentation, there is this:
`OnePub Subscription requirements OnePub offers a special licensing model for distributing packages to Customers. You can add a Distribution license to your Plan which provides for a usage based model (as opposed to the normal 'per seat' licenses). There are a no. of requirements around the Distribution license. user internal to your organisation, must have a 'per seat' license. external users must use a company domain email (e.g. me@onepub.dev vs me@gmail.com) external users are restricted to read only access. If you want to use Teams to restrict external users to specific packages then you will need a teams license.`

So I have a private flutter package that I want to distribute to my customers. After reading the above description, I still have no idea what I should to do give them the read-only access to my flutter package in OnePub. I'm confused.

- Should I invite my customer by "Member Invitation"? 
- Do they count towards my no. of members? 
- How do I add "Distribution License"?
- If my customer has a group of developers, can they share the same Distribution license? 
- Is "External user" actually a "Member"
- How can I view "Distribution License" in my OnePub portal?
- I only have a customer, and the customer has a group of developers, what OnePub plan should I go for? 
- How many members should I be creating, excluding myself?
- What is "PackageLicensee"?
![image](https://user-images.githubusercontent.com/126567837/221835972-78e3792d-7a68-4e6e-8557-21d545362d25.png)

**Comment by jeremy-carbonne** [comment]
I'm in the same situation.
I will share a private git repo to my customer 😢

Your documentation is great, but several things do not seem to be available yet, the CI role for example is not visible

**Comment by bsutton** [comment, marked as answer]
Sorry for the late response on this one, we are in the final days of moving out of beta so it's been a bit hectic, to say the least.

If the final testing doesn't reveal any issues we will be moving out of Beta on Monday.

Firstly to the simple issues:

@jeremy-carbonne 

The CICD role does exist but there is a known bug that results in the Role not being visible.
We have a fix for this and if all goes well it will be rolled by the end of next week.
In the meant time, you can just create another account with the Member role and achieve most of the core objectives (the CICD role is more restricted than a normal Member, otherwise there is no difference).

As to the distribution question.

So the documentation isn't quite right, I'm in the process of updating it and it should be ready next week.

So the requirements are:
* You do need a Team licence but this only applies to members within your organisation.
* YOu need a Distribution license, currently have to activate this from the backend (this will change in the future).
* All your internal team members must share an email domain name
* You must purchase a Seat for each of your internal team members
* External users are members, they just don't require a per seat licence.
* External users must have the 'Customer' role added to them.

We haven't actually released the Distribution License publicly but it will be live by the end of the week (10th of March).

We would be rather keen to have you both on board as the first live users.
We understand there would be a some risk for your organisations so are prepared to offer a few benefits.
1) a 2 month free trial of the distribution feature.
2) 50% of the distribution license price for the first 12 months post the trial
3) provide you with my direct contact details
4) next business day turn around on any bugs you might uncover.

If you are interested post a response and I will get you set up.

Brett

**Reply by isaac-gofa** [reply]
Interested. Thanks for your help!

And I would like to pay, just it's still unclear about the specific costs.
Such as Team license cost, and the distribution license cost.
And cost increase if I have more developers or more customers.
Right now I'm the only developer to manage the private repo, and I only have one customer to distribute the private repo to.

**Reply by isaac-gofa** [reply]
Hello @bsutton, any updates?

**Comment by bsutton** [comment]
A couple of missing points

You can add any no. of customers.
Each customer developer should be given their own account.

**Comment by bsutton** [comment]
- Should I invite my customer by "Member Invitation"?

Yes, on Monday the will be an updated member invite that lets you invite a
member to a specific team. Create a team Pet customer

   - Do they count towards my no. of members?

No.

   - How do I add "Distribution License"?

We do that manually

   - If my customer has a group of developers, can they share the same
   Distribution license?

Only  single distribution licence for all your customers

   - Is "External user" actually a "Member"

Yes

   - How can I view "Distribution License" in my OnePub portal?

There is a new product tab that shows your licence. Coming Monday

**Comment by bsutton** [comment]
Customer are unlimited.

Developers of customers are unlimited.

You need at a team or enterprise seat licence for each of the developers
that work for your company.

**Comment by bsutton** [comment]
We have a release planned for tomorrow which include the required changes.

Will update you once it has been pushed.

**Comment by bsutton** [comment]
Just thought I should give you an update.

1)  distribution license 
Mostly done but we found that the permission model went deeper than expected so we are going to need until next (mid -late) week to iron out the last few problems. After CICD this is currently our highest priority.
2) CICD
The work on this is complete but we need to run a few more tests. I'm optimistic that this will be released tomorrow.
3) delete a member
There is a new menu item on the Member view 'Remove Member'. This correctly removes a member from your organisation.
This replaced the previous 'Delete' item, which conveyed the wrong semantics (we are just detaching the member from your organisation as opposed to actually deleting them as they may belong to other organisations).
Otherwise this is ready to go.

I will keep you posted as we progress.

**Comment by bsutton** [comment]
@isaac-gofa @jeremy-carbonne 

Just letting you know that we have finally released the distribution license and reworked the CI/CD model.

To enable the distribution license, you start by Inviting a Customer from the 'Invite' link on the home page.

We have also updated the doco:

CICD
https://docs.onepub.dev/guides/ci-cd

Distribution license
https://docs.onepub.dev/workflows/distributing-to-customers

Any issues, please let me know.
